### ISSUE TITLE
Cartpole notebook results don't hold any more

### ISSUE THREAD

**Issue by shwang** [HumanCompatibleAI/imitation#3]
When the notebook was generated:
![image](https://user-images.githubusercontent.com/1750835/54394404-01ab6c00-466a-11e9-9f7b-7d79851d3a8f.png)

(edit: what I remember is that the average episode reward floated around 500 for many more training epochs, and then collapsed)

Now:
![image](https://user-images.githubusercontent.com/1750835/54394439-1a1b8680-466a-11e9-9802-55ba60978868.png)
![image](https://user-images.githubusercontent.com/1750835/54394435-1556d280-466a-11e9-9fd0-24784754cbe3.png)

I should explore whether notebook still replicates after I checkout a commit around Dec 20 when I first got these results (say, f9669e6ee3).

**Comment by shwang** [comment]
Two runs using the new `cartpole_airl_best.gin` settings & the new replay buffer:

Plot Formats:
  * tracking discrimnator loss over time
  * Tracking average true cartpole performance of different policies at each training epoch (only generator policy is trained, other policies are fixed). The maximum episode reward is 499, and there is 1 reward for each timestep alive.
  * Tracking the average AIRL episode reward collected by each policy at each training epoch.

## Run 1:
![image](https://user-images.githubusercontent.com/1750835/60547875-add69880-9cd5-11e9-92a8-3f4339522b05.png)
![image](https://user-images.githubusercontent.com/1750835/60547884-b0d18900-9cd5-11e9-9ebe-9d85afe79f0d.png)
![image](https://user-images.githubusercontent.com/1750835/60547887-b333e300-9cd5-11e9-89f0-7d54c0601780.png)



## Run 2:
![image](https://user-images.githubusercontent.com/1750835/60547899-b8912d80-9cd5-11e9-8b7a-b01f0b570583.png)
![image](https://user-images.githubusercontent.com/1750835/60547904-baf38780-9cd5-11e9-8b73-7534d74aba90.png)
![image](https://user-images.githubusercontent.com/1750835/60547912-bdee7800-9cd5-11e9-8a9d-ea95d675b5ed.png)

**Comment by AdamGleave** [comment]
These are interesting plots, thanks for sharing.

My hunch is something wrong with the discriminator, or possibly just not trained for long enough -- discriminator loss seems to keep going down.

Note in run #1 the generated policy gets similar *learned* reward to the expert, and in run #2 outperforms it. So the RL step is probably doing OK. But it can get good learned reward without actually replicating expert behavior.

Run #2 is particularly odd. The discriminator loss spikes substantially. I can't tell from the plot whether this is before or after the generator's true reward performance drops. This is worth checking if you still have the data. If it's after, then this suggests that: discriminator broke => generator starts getting crazy reward => generator performance plummets. If before, then maybe it's something more to do with instability of adversarial training.

Have we tried just decreasing discriminator learning rate and/or boosting the batch size, and training for longer?

**Comment by AdamGleave** [comment]
I think this can be closed now #37 is merged
